File: AtlasWeb/Core/Modules/Saglik_Lojistik/Stok_Modulleri/Sihhiye_Modulleri/Tasinir_Mal_Islem_Modulu/ChattelDocumentInputWithAccountancyNewFormViewModel.ts
//$702FBACE
import { BaseViewModel } from "NebulaClient/Model/BaseViewModel";
import { ChattelDocumentInputWithAccountancy, PTSStockActionDetail } from 'NebulaClient/Model/AtlasClientModel';
import { ChattelDocumentInputDetailWithAccountancy } from 'NebulaClient/Model/AtlasClientModel';
import { StockActionSignDetail } from 'NebulaClient/Model/AtlasClientModel';
import { BudgetTypeDefinition } from 'NebulaClient/Model/AtlasClientModel';
import { Store } from 'NebulaClient/Model/AtlasClientModel';
import { Accountancy } from 'NebulaClient/Model/AtlasClientModel';
import { Material } from 'NebulaClient/Model/AtlasClientModel';
import { StockCard } from 'NebulaClient/Model/AtlasClientModel';
import { DistributionTypeDefinition } from 'NebulaClient/Model/AtlasClientModel';
import { StockLevelType } from 'NebulaClient/Model/AtlasClientModel';
import { ResUser, Supplier } from 'NebulaClient/Model/AtlasClientModel';
import { Type } from 'NebulaClient/ClassTransformer';
import { Guid } from "NebulaClient/Mscorlib/Guid";
import { PTSMaterial } from "./BaseChattelDocumentWithPurchaseForm";


export class ChattelDocumentInputWithAccountancyNewFormViewModel extends BaseViewModel {
    @Type(() => ChattelDocumentInputWithAccountancy)
    public _ChattelDocumentInputWithAccountancy: ChattelDocumentInputWithAccountancy = new ChattelDocumentInputWithAccountancy();
    @Type(() => ChattelDocumentInputDetailWithAccountancy)
    public ChattelDocumentDetailsWithAccountancyGridList: Array<ChattelDocumentInputDetailWithAccountancy> = new Array<ChattelDocumentInputDetailWithAccountancy>();
    @Type(() => StockActionSignDetail)
    public StockActionSignDetailsGridList: Array<StockActionSignDetail> = new Array<StockActionSignDetail>();
    @Type(() => BudgetTypeDefinition)
    public BudgetTypeDefinitions: Array<BudgetTypeDefinition> = new Array<BudgetTypeDefinition>();
    @Type(() => Store)
    public Stores: Array<Store> = new Array<Store>();
    @Type(() => Accountancy)
    public Accountancys: Array<Accountancy> = new Array<Accountancy>();
    @Type(() => Material)
    public Materials: Array<Material> = new Array<Material>();
    @Type(() => StockCard)
    public StockCards: Array<StockCard> = new Array<StockCard>();
    @Type(() => DistributionTypeDefinition)
    public DistributionTypeDefinitions: Array<DistributionTypeDefinition> = new Array<DistributionTypeDefinition>();
    @Type(() => StockLevelType)
    public StockLevelTypes: Array<StockLevelType> = new Array<StockLevelType>();
    @Type(() => ResUser)
    public ResUsers: Array<ResUser> = new Array<ResUser>();
    @Type(() => Supplier)
    public Suppliers: Array<Supplier> = new Array<Supplier>();

    public chattelDocumentInputWithAccountancyGridViewModel: Array<ChattelDocumentInputWithAccountancyGridViewModel> = new Array<ChattelDocumentInputWithAccountancyGridViewModel>();
    @Type(() => PTSStockActionDetail)
    public PTSStockActionDetails: Array<PTSStockActionDetail> = new Array<PTSStockActionDetail>();
    @Type(() => PTSMaterial)
    public PTSMaterials: Array<PTSMaterial> = new Array<PTSMaterial>();

}

export class ChattelDocumentInputWithAccountancyGridViewModel {
    public MaterialObjectID: Guid;
    public MaterialName: string;
    public Material: Material;
    public Barcode: string;
    public UnitType: string;
    public Amount: number;
    public NotDiscountedUnitPrice: number;
    public KDV: number;
    public DiscountRate: number;
    //İndirim ve Kdv Oranı ile hesaplanan birim fiyat
    public CalculatedUnitPriceOfMaterial: number;
    public TotalPriceNotDiscount: number;
    public TotalDiscountAmount: number;
    public TotalPrice: number;
    public LotNo: string;
    public ExpirationDate: Date;
    public StockLevelType: string;
    public StockLevelTypeObjectID: Guid;
    public RetrievalYear: number;
    public ProductionDate: Date;
}

File: AtlasWeb/Core/Modules/Saglik_Lojistik/Stok_Modulleri/Sihhiye_Modulleri/Tasinir_Mal_Islem_Modulu/ChattelDocumentInputWithAccountancyNewFormViewModel.cs
//$BA78D070
using System;
using System.Linq;
using TTInstanceManagement;
using Core.Models;
using TTObjectClasses;
using TTDefinitionManagement;
using TTUtils;
using System.Collections.Generic;
using TTDataDictionary;

namespace Core.Controllers
{
    public partial class ChattelDocumentInputWithAccountancyServiceController
    {
        partial void PostScript_ChattelDocumentInputWithAccountancyNewForm(ChattelDocumentInputWithAccountancyNewFormViewModel viewModel, ChattelDocumentInputWithAccountancy chattelDocumentInputWithAccountancy, TTObjectStateTransitionDef transDef, TTObjectContext objectContext)
        {

        }

        partial void AfterContextSaveScript_ChattelDocumentInputWithAccountancyNewForm(ChattelDocumentInputWithAccountancyNewFormViewModel viewModel, ChattelDocumentInputWithAccountancy chattelDocumentInputWithAccountancy, TTDefinitionManagement.TTObjectStateTransitionDef transDef, TTObjectContext objectContext)
        {
            try
            {
                if (viewModel._ChattelDocumentInputWithAccountancy.CurrentStateDefID == ChattelDocumentInputWithAccountancy.States.Completed)
                {
                    if (TTObjectClasses.SystemParameter.GetParameterValue("SENDTOMKYSFORAFTERCONTEXT", "TRUE") == "TRUE")
                    {
                        var sonucMesaji = chattelDocumentInputWithAccountancy.SendMKYSForInputDocument(Common.CurrentResource.MkysPassword);
                        TTUtils.InfoMessageService.Instance.ShowMessage(sonucMesaji);
                    }

                    if (chattelDocumentInputWithAccountancy.InPatientPhysicianApplication != null)
                    {
                        foreach (ChattelDocumentInputDetailWithAccountancy detail in chattelDocumentInputWithAccountancy.ChattelDocumentInputDetailsWithAccountancy)
                        {
                            BaseTreatmentMaterial baseTreatmentMaterial = new BaseTreatmentMaterial(objectContext);
                            baseTreatmentMaterial.Material = detail.Material;
                            baseTreatmentMaterial.Amount = detail.Amount;
                            baseTreatmentMaterial.Patient = detail.Patient;
                            baseTreatmentMaterial.Store = chattelDocumentInputWithAccountancy.Store;
                            DateTime now = Common.RecTime();
                            DateTime actionDate = chattelDocumentInputWithAccountancy.WaybillDate.Value.AddHours(now.Hour).AddMinutes(now.Minute).AddSeconds(now.Second);
                            baseTreatmentMaterial.ActionDate = actionDate;
                            if (chattelDocumentInputWithAccountancy.InPatientPhysicianApplication.SubEpisode.ClosingDate < now)
                                baseTreatmentMaterial.PricingDate = chattelDocumentInputWithAccountancy.InPatientPhysicianApplication.SubEpisode.ClosingDate;
                            chattelDocumentInputWithAccountancy.InPatientPhysicianApplication.TreatmentMaterials.Add(baseTreatmentMaterial);
                        }
                    }
                    objectContext.Save();


                    if (chattelDocumentInputWithAccountancy.IsPTSAction == true)
                        chattelDocumentInputWithAccountancy.SendITSReceiptNotification();
                }
            }
            catch (Exception ex)
            {
                throw;
            }
        }
    }
}

namespace Core.Models
{
    public partial class ChattelDocumentInputWithAccountancyNewFormViewModel
    {
        public List<ChattelDocumentInputWithAccountancyGridViewModel> chattelDocumentInputWithAccountancyGridViewModel = new List<ChattelDocumentInputWithAccountancyGridViewModel>();
    }

    public class ChattelDocumentInputWithAccountancyGridViewModel
    {
        public Material Material;
        public Guid MaterialObjectID;
        public string MaterialName;
        public string Barcode;
        public string UnitType;
        public int Amount;
        public double NotDiscountedUnitPrice;
        public long KDV;
        public double DiscountRate;
        public double CalculatedUnitPriceOfMaterial;
        public double TotalPriceNotDiscount;
        public double TotalDiscountAmount;
        public double TotalPrice;
        public string LotNo;
        public DateTime ExpirationDate;
        public string StockLevelType;
        public Guid StockLevelTypeObjectID;
        public int RetrievalYear;
        public DateTime ProductionDate;
    }


}